kanidmd/idm/src/idm/credupdatesession.rs

```diff
@@ -0,0 +1,1696 @@
+            OperationError::InvalidState
+        })?;
+
+        trace!(?session);
+
+        let mut modlist = ModifyList::new();
+
+        // Setup mods for the various bits. We always assert an *exact* state.
+
+        match &session.primary {
+            Some(ncred) => {
+                modlist.push_mod(Modify::Purged(AttrString::from("primary_credential")));
+                let vcred = Value::new_credential("primary", ncred.clone());
+                modlist.push_mod(Modify::Present(
+                    AttrString::from("primary_credential"),
+                    vcred,
+                ));
+            }
+            None => {
+                modlist.push_mod(Modify::Purged(AttrString::from("primary_credential")));
+            }
+        };
+
+        // If an intent token was used, remove it's former value, and add it as consumed.
+        if let Some(intent_token_id) = session.intent_token_id {
+            modlist.push_mod(Modify::Removed(
+                AttrString::from("credential_update_intent_token"),
+                PartialValue::IntentToken(intent_token_id),
+            ));
+            modlist.push_mod(Modify::Present(
+                AttrString::from("credential_update_intent_token"),
+                Value::IntentToken(intent_token_id, IntentTokenState::Consumed),
+            ));
+        };
+
+        // Are any other checks needed?
+
+        // Apply to the account!
+        trace!(?modlist, "processing change");
+
+        self.qs_write
+            .internal_modify(
+                // Filter as executed
+                &filter!(f_eq("uuid", PartialValue::new_uuidr(&session.account.uuid))),
+                &modlist,
+            )
+            .map_err(|e| {
+                request_error!(error = ?e);
+                e
+            })
+    }
+}
+
+impl<'a> IdmServerCredUpdateTransaction<'a> {
+    fn get_current_session(
+        &self,
+        cust: &CredentialUpdateSessionToken,
+        ct: Duration,
+    ) -> Result<CredentialUpdateSessionMutex, OperationError> {
+        let session_token: CredentialUpdateSessionTokenInner = self
+            .token_enc_key
+            .decrypt_at_time(&cust.token_enc, None, ct.as_secs())
+            .map_err(|e| {
+                admin_error!(?e, "Failed to decrypt credential update session request");
+                OperationError::SessionExpired
+            })
+            .and_then(|data| {
+                serde_json::from_slice(&data).map_err(|e| {
+                    admin_error!(err = ?e, "Failed to deserialise credential update session request");
+                    OperationError::SerdeJsonError
+                })
+            })?;
+
+        // Check the TTL
+        // Asserted by the above.
+        /*
+        if ct >= token.max_ttl {
+            trace!(?ct, ?token.max_ttl);
+            security_info!(%token.sessionid, "session expired");
+            return Err(OperationError::SessionExpired);
+        }
+        */
```

Comment:
Well, the question is do we rely on the fernet time checks, or should we do them ourself? I think it could be clearer here to do them ourself, but it adds risk to us to ... well, do them.